server/src/main/java/io/crate/planner/optimizer/rule/SwapHashJoin.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.planner.optimizer.rule;
+
+import static io.crate.planner.optimizer.matcher.Pattern.typeOf;
+
+import java.util.function.Function;
+
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.TransactionContext;
+import io.crate.planner.operators.Eval;
+import io.crate.planner.operators.HashJoin;
+import io.crate.planner.operators.LogicalPlan;
+import io.crate.planner.optimizer.Rule;
+import io.crate.planner.optimizer.costs.PlanStats;
+import io.crate.planner.optimizer.matcher.Captures;
+import io.crate.planner.optimizer.matcher.Pattern;
+
+public class SwapHashJoin implements Rule<HashJoin> {
+
+    private final Pattern<HashJoin> pattern = typeOf(HashJoin.class)
+        .with(j -> j.isSwapSidesDone() == false);
+
+    @Override
+    public Pattern<HashJoin> pattern() {
+        return pattern;
+    }
+
+    @Override
+    public LogicalPlan apply(HashJoin plan,
+                             Captures captures,
+                             PlanStats planStats,
+                             TransactionContext txnCtx,
+                             NodeContext nodeCtx,
+                             Function<LogicalPlan, LogicalPlan> resolvePlan) {
+        var lhStats = planStats.get(plan.lhs());
+        var rhStats = planStats.get(plan.rhs());
+        boolean expectedRowsAvailable = lhStats.numDocs() != -1 && rhStats.numDocs() != -1;
+        // We move the smaller table to the right side since benchmarking
+        // revealed that this improves performance in most cases.
+        if (expectedRowsAvailable && lhStats.numDocs() < rhStats.numDocs() &&
+            lhStats.sizeInBytes() <= rhStats.sizeInBytes()) {
```

Comment:
I'd vote to keep this PR "clean" with only introducing a proper rule for the swapping and later on decide on this additional check after properly benchmarking it.

Reply:
Please have a look here https://github.com/crate/crate/pull/6985, it seems that decision has to do with the number of repetitions (= number of rows for right table) and doesn't involve sizes, and the benchmarks showed that even if (because of size) the right would fit into fewer blocks (as a left table) the performance diff is minimal.

Reply:
https://github.com/crate/crate/issues/13501 is derived from a customer issue with exactly that issue. I am ok with leaving this out and do some benchmarks.

Reply:
I'd say yes please, it's better to have concrete metrics before deciding.